### PR TITLE
Preserve per-turn metrics and add per-turn dashboard CSV rows

### PR DESCRIPTION
**Pull request by aas008**
## Summary

- **`run_guidellm_benchmark/main.py`**: `trim_benchmark_json` now strips only bulky text fields (`request_args`, `output`, `reasoning_output` — ~50-100KB each) instead of deleting the entire `requests` field. Preserves per-request metrics (~800 bytes each: TTFT, ITL, TPOT, latency, token counts, turn_index, conversation_id).

- **`csv_export.py`**: For multi-turn benchmarks, emits additional rows per turn alongside the existing aggregate row. Example: concurrency=1 with 3 turns → 4 CSV rows (1 aggregate + 3 per-turn). New `turn` column is empty on aggregate rows, `0`/`1`/`2` on per-turn rows. Also populates the existing `turns`, `prefix_tokens`, `prefix_count` columns (were always empty). Non-multi-turn benchmarks are unaffected.

## Grafana usage

Filter by `turn` column:
- Empty → aggregate row (existing behavior)
- Non-empty → per-turn breakdown (compare TTFT/ITL/latency across turns)

## Test plan

- [x] Verified with multi-turn benchmarks.json: 3 rates × 3 turns → 12 rows (3 aggregate + 9 per-turn)
- [x] Verified non-multi-turn produces same rows as before (no per-turn rows emitted)
- [x] Ruff lint + format pass
- [x] TTFT growth across turns confirmed (69.8ms → 74.1ms → 75.4ms at conc=1)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - CSV exports now include turn information, prefix token counts, and prefix counts.
  - Multi-turn benchmarks now produce per-turn rows with request statistics and percentile metrics.
  - Benchmark summaries retain per-turn metadata, token counts, and computed metrics.

- **Bug Fixes**
  - Benchmark JSON trimming now preserves request entries while removing only request arguments and generated content fields.

### CODE REVIEW

**Comment by aas008** [comment]
/test fournos rhaiis nvidia
/pipeline forge-full
/cluster hera
/exclusive false
/var tests.rhaiis.model_key: qwen3-0_6b
/var tests.rhaiis.workload_key: custom
/var tests.rhaiis.run_benchmark: true
/var tests.rhaiis.warmup: false
/var tests.rhaiis.version: "multi-turn-quick-test"
/var rhaiis.engine: vllm
/var rhaiis.profiler.enabled: false
/var caliper.postprocess.csv_dashboard.enabled: false
/var rhaiis.agent_analysis.enabled: false
/var tests.rhaiis.slack_notify_always: true
/var workloads.custom.data: "prompt_tokens=128,output_tokens=128,turns=3,prefix_tokens=10000,prefix_count=512"
/var workloads.custom.rates: [1, 10, 50]
/var workloads.custom.max_seconds: 60

**Comment by aas008** [comment]
/test fournos rhaiis nvidia
/pipeline forge-full
/cluster hera
/exclusive false
/var tests.rhaiis.model_key: qwen3-0_6b
/var tests.rhaiis.workload_key: custom
/var tests.rhaiis.run_benchmark: true
/var tests.rhaiis.warmup: false
/var tests.rhaiis.version: "multi-turn-quick-test"
/var rhaiis.engine: vllm
/var rhaiis.profiler.enabled: false
/var caliper.postprocess.csv_dashboard.enabled: false
/var rhaiis.agent_analysis.enabled: false
/var tests.rhaiis.slack_notify_always: true
/var workloads.custom.data: "prompt_tokens=128,output_tokens=128,turns=3,prefix_tokens=10000,prefix_count=512"
/var workloads.custom.rates: [1, 10, 50]
/var workloads.custom.max_seconds: 60

**Comment by kpouget** [comment]
thanks,
/lgtm

**Comment by aas008** [comment]
/test fournos rhaiis nvidia
/pipeline forge-full
/cluster hera
/exclusive false
/var tests.rhaiis.model_key: qwen3-0_6b
/var tests.rhaiis.workload_key: custom
/var tests.rhaiis.run_benchmark: true
/var tests.rhaiis.warmup: false
/var tests.rhaiis.version: "multi-turn-quick-test"
/var rhaiis.engine: vllm
/var rhaiis.profiler.enabled: false
/var caliper.postprocess.csv_dashboard.enabled: false
/var rhaiis.agent_analysis.enabled: false
/var tests.rhaiis.slack_notify_always: true
/var workloads.custom.data: "prompt_tokens=128,output_tokens=128,turns=3,prefix_tokens=10000,prefix_count=512"
/var workloads.custom.rates: [1, 10, 50]
/var workloads.custom.max_seconds: 60

**Comment by aas008** [comment]
/test fournos rhaiis nvidia
/pipeline forge-full
/cluster hera
/exclusive false
/var tests.rhaiis.model_key: qwen3-0_6b
/var tests.rhaiis.workload_key: custom
/var tests.rhaiis.run_benchmark: true
/var tests.rhaiis.warmup: false
/var tests.rhaiis.version: "multi-turn-quick-test"
/var rhaiis.engine: vllm
/var rhaiis.profiler.enabled: false
/var caliper.postprocess.csv_dashboard.enabled: false
/var rhaiis.agent_analysis.enabled: false
/var tests.rhaiis.slack_notify_always: true
/var workloads.custom.data: "prompt_tokens=128,output_tokens=128,turns=3,prefix_tokens=10000,prefix_count=512"
/var workloads.custom.rates: [1, 10, 50]
/var workloads.custom.max_seconds: 60

**Comment by aas008** [comment]
## Code Review

**Decision: `ready`**

> **Scope**: Reviewed 9 unique commits out of 9 total. Target branch: `main`.

### Intent & Optimality (pr-intent-review)
**Understood intent**: Adds per-turn metric breakdowns for multi-turn GuideLLM benchmarks. Previously, one CSV row per concurrency rate; now, additional rows per turn (`turn_index=0,1,2...`) alongside aggregate rows (`turn_index=empty`). Two coordinated changes: (1) preserve request-level data in trimmed benchmark JSON (selective field stripping instead of removing entire `requests` key), (2) generate per-turn CSV rows through the production caliper pipeline.

**Optimality**: Optimal — flows through the production pipeline (`_extract_dashboard_metrics` → `compute_dashboard_kpis` → `write_dashboard_csv`) rather than a parallel code path. Clean backward compatibility: single-turn benchmarks produce no per-turn rows via the `len(all_turns) <= 1` guard. Column naming (`turn_index` vs `turns`) avoids confusion.

### Breaking Changes (code-review-breaking-change)
Two medium-severity changes identified:

1. **CSV column reordering**: `RHAIIS_FIELDNAMES` adds `turn_index` at position 8 and moves `turns` from position ~55 to position 9. Name-based readers (including the Streamlit dashboard) are unaffected. Positional readers would silently get wrong data — verified that `DictWriter` with `extrasaction="ignore"` prevents crashes.

2. **`trim_benchmark_json` behavioral change**: No longer strips the `requests` field entirely; selectively removes `request_args`, `output`, `reasoning_output`. Trimmed benchmark JSON files will be larger (~800 bytes per request retained vs near-zero before). No external callers (all 3 call sites in `main.py`).

3. **Additional CSV rows**: Multi-turn benchmarks now emit per-turn rows with `__turn_N` run_id suffix. Aggregate rows preserved. Companion PR [#101](https://github.com/openshift-psap/performance-dashboard/pull/101) handles the Streamlit dashboard filtering.

### Change Size (code-review-change-size)
**PASS**: 531 additions, 7 deletions across 9 files (threshold: 800). Largest file: `dashboard.py` (178 lines, threshold: 500). Test file: 320 lines (61% of additions). Two tightly coupled logical units with clear dependency. No split recommended.

### Test Integrity (code-review-test-integrity)
No test deletions, CI tampering, or coverage suppression. 24 tests added in `test_per_turn_metrics.py` covering `_percentile`, `_extract_per_turn_curves` (including errored request discovery and JSON round-trip safety), `_turn_metric_values` (including errored count propagation), and `trim_benchmark_json`.

### Comment Density (code-review-comment-density)
No comment density issues found.

### Race Condition (code-review-race-condition-check)
No race condition issues found.

---

### Action Items

- [x] ~~**1. [MEDIUM] Schema drift: guidellm and llm_d fieldnames missing turn_index/turns** — `guidellm/csv_dashboard.py`, `llm_d/csv_dashboard.py`~~ **RESOLVED** in commit `ac7d6e8`
<details>
<summary>Details</summary>

`BASIC_GUIDELLM_FIELDNAMES` and llm_d `DASHBOARD_FIELDNAMES` both lacked `turn_index` and `turns`. When multi-turn benchmarks run through these pipelines, per-turn rows would appear without a turn identifier (`extrasaction="ignore"` silently drops unmapped columns). llm_d confirmed to run multi-turn benchmarks. Fixed by adding columns and mapper entries to both plugins. Outside spec scope.
</details>

- [x] ~~**2. [LOW] Integer dict keys through JSON cache** — `dashboard.py:_extract_per_turn_curves`~~ **RESOLVED** in commit `339577e`
<details>
<summary>Details</summary>

`_extract_per_turn_curves` now returns `dict[str, ...]` with stringified integer keys so they survive JSON round-trips through the caliper cache (`json.dumps` converts int keys to strings). Consumer in `compute_dashboard_kpis` sorts numerically via `key=lambda x: int(x[0])`. Outside spec scope.
</details>

- [x] ~~**3. [LOW] Missing test coverage for per-turn logic** — `dashboard.py`, `main.py`~~ **RESOLVED** in commit `339577e`
<details>
<summary>Details</summary>

Added `test_per_turn_metrics.py` with 24 tests covering `_percentile` (edge cases, interpolation), `_extract_per_turn_curves` (multi-turn string keys, JSON round-trip, missing turns, errored request discovery), `_turn_metric_values` (basic stats, empty/None handling, errored count propagation), and `trim_benchmark_json` / `_trim_request_entry` (selective stripping, recursive traversal). Outside spec scope.
</details>

- [x] ~~**4. [MEDIUM] Errored requests not included in per-turn accounting** — `dashboard.py:_extract_per_turn_curves`~~ **RESOLVED** in commit `41f585d`
<details>
<summary>Details</summary>

Turn discovery now scans both `requests.successful` and `requests.errored` so turns with only failures are still visible. Per-turn error count is passed through to `_turn_metric_values` instead of hardcoding zero. Tests cover both errored-counted and errored-only-turn-discovered scenarios. Reported by CodeRabbit.
</details>

- [ ] **5. [INFO] Artifact size increase** — `main.py:trim_benchmark_json`
<details>
<summary>Details</summary>

Preserving `requests` (minus bulk fields) adds ~800 bytes per request. For Profile6 runs (12,800 requests), ~10MB per run vs near-zero before. Outside spec scope.
</details>

---
— Reviewed by Coding Agent

File: projects/rhaiis/postprocess/csv_export.py [hunk 1]
```diff
@@ -277,10 +291,148 @@ def _total_stat(metric_name: str, stat: str):
...
+    if not successful_reqs:
+        return []
+
+    by_turn: dict[int, list[dict]] = defaultdict(list)
+    for req in successful_reqs:
+        turn = req.get("info", {}).get("turn_index", 0)
+        by_turn[turn].append(req)
+
+    if len(by_turn) <= 1:
+        return []
```
[the remark this answers is not in the document]
Comment (aas008) [reply projects/rhaiis/postprocess/csv_export.py]: This comment references `csv_export.py:_per_turn_rows` which was removed in a prior rebase. The per-turn logic now lives in `dashboard.py:_extract_per_turn_curves`. No action needed — this file/function no longer exists.

File: projects/guidellm/postprocess/guidellm/dashboard.py [hunk 2]
```diff
@@ -351,9 +351,129 @@ def _extract_dashboard_metrics(node: TestBaseNode) -> tuple[dict[str, Any], dict
...
+    when multiple turns exist, otherwise an empty dict.  Keys are
+    stringified integers so they survive JSON round-trips through the
+    caliper cache.
+    """
+    all_turns: set[int] = set()
+    for bench in benchmarks:
+        for req in bench.get("requests", {}).get("successful", []):
+            ti = req.get("info", {}).get("turn_index")
+            if ti is not None:
+                all_turns.add(ti)
```
[the remark this answers is not in the document]
Comment (aas008) [reply projects/guidellm/postprocess/guidellm/dashboard.py]: Fixed in 41f585d3 — `_extract_per_turn_curves` now discovers turn indices from both `successful` and `errored` request buckets, and passes the per-turn error count through to `_turn_metric_values` instead of hardcoding zero. Tests added for both cases.

**Comment by aas008** [comment]
## Babysit PR — Status

### Current State
- **SHA**: `41f585d3`
- **CI**: 5/6 passed (tide pending)
- **Mergeability**: MERGEABLE
- **Review**: All code review items resolved

### ⚠️ Blocker: Missing `lgtm` label
The `lgtm` label was added by kpouget on Sep 16 but automatically removed when new commits were pushed (Sep 16–21). Prow/tide requires the `lgtm` label to merge.

**Action needed**: A reviewer needs to re-post `/lgtm` on the latest SHA (`41f585d3`).

### Resolved review items

| # | Finding | Commit | Status |
|---|---------|--------|--------|
| 1 | Schema drift: guidellm/llm_d fieldnames missing turn_index/turns | `ac7d6e8` | Resolved |
| 2 | Integer dict keys through JSON cache | `339577e` | Resolved |
| 3 | Missing test coverage for per-turn logic | `339577e` | Resolved |
| 4 | Errored requests not in per-turn accounting | `41f585d` | Resolved |

### Open items

| # | Finding | Status |
|---|---------|--------|
| 5 | Artifact size increase (~10MB for Profile6 runs) | INFO — no action needed |

### Notes
- Stashed uncommitted changes to `projects/core/library/export.py` before monitoring.
- Label timeline: `lgtm` added 2026-09-16T07:27:02Z, removed 2026-09-16T22:07:24Z (new commits pushed).
- Monitoring continues for new review comments and label changes.

---
— Bazinga PR Babysitter

**Comment by aas008** [comment]
/test fournos rhaiis nvidia
/pipeline forge-full
/cluster hera
/exclusive false
/var tests.rhaiis.model_key: qwen3-0_6b
/var tests.rhaiis.workload_key: custom
/var tests.rhaiis.run_benchmark: true
/var tests.rhaiis.warmup: false
/var tests.rhaiis.version: "multi-turn-quick-test"
/var rhaiis.engine: vllm
/var rhaiis.profiler.enabled: false
/var caliper.postprocess.csv_dashboard.enabled: false
/var rhaiis.agent_analysis.enabled: false
/var tests.rhaiis.slack_notify_always: true
/var workloads.custom.data: "prompt_tokens=128,output_tokens=128,turns=3,prefix_tokens=10000,prefix_count=512"
/var workloads.custom.rates: [1, 10, 50]
/var workloads.custom.max_seconds: 60
